### PR TITLE
feat(number): add multipleOf to faker.number.int

### LINKED ISSUE #2584: add multipleOf to faker.number.int
_Originally posted by @matthewmayer in https://github.com/faker-js/faker/pull/2581#discussion_r1431611815_
            
Similar to faker.number.float, faker.number.int should also support a multipleOf parameter

### PR DESCRIPTION
**Pull request by matthewmayer**
fix #2584

### CODE REVIEW

File: src/modules/number/index.ts [hunk 1]
```diff
@@ -23,9 +23,11 @@ export class NumberModule extends SimpleModuleBase {
    * @param options Maximum value or options object.
    * @param options.min Lower bound for generated number. Defaults to `0`.
    * @param options.max Upper bound for generated number. Defaults to `Number.MAX_SAFE_INTEGER`.
+   * @param options.multipleOf Generated number must be a multiple of the given integer. Defaults to `1`.
    *
    * @throws When `min` is greater than `max`.
-   * @throws When there are no integers between `min` and `max`.
+   * @throws When there are no suitable integers between `min` and `max`.
+   * @throws When multipleOf is not a positive integer.
```
Comment (ST-DDT) [inline src/modules/number/index.ts]:
```suggestion
   * @throws When `multipleOf` is not a positive integer.
```

**Review by ST-DDT — DISMISSED** [review]

File: src/modules/number/index.ts [hunk 2]
```diff
@@ -54,13 +57,17 @@ export class NumberModule extends SimpleModuleBase {
            * @default Number.MAX_SAFE_INTEGER
            */
           max?: number;
+          /**
+           * Generated number must be a multiple of the given integer. Defaults to `1`.
```
Comment (ST-DDT) [inline src/modules/number/index.ts]:
```suggestion
           * Generated number must be a multiple of the given integer.
           *
           * @default 1
```

File: src/modules/number/index.ts [hunk 3]
```diff
@@ -23,9 +23,11 @@ export class NumberModule extends SimpleModuleBase {
    * @param options Maximum value or options object.
    * @param options.min Lower bound for generated number. Defaults to `0`.
    * @param options.max Upper bound for generated number. Defaults to `Number.MAX_SAFE_INTEGER`.
+   * @param options.multipleOf Generated number must be a multiple of the given integer. Defaults to `1`.
```
Comment (ST-DDT) [inline src/modules/number/index.ts]:
```suggestion
   * @param options.multipleOf Generated number will be a multiple of the given integer. Defaults to `1`.
```

File: src/modules/number/index.ts [hunk 4]
```diff
@@ -78,9 +87,30 @@ export class NumberModule extends SimpleModuleBase {
...
+    if (multipleOf <= 0) {
+      throw new FakerError(`multipleOf should be greater than 0.`);
+    }
+
     // @ts-expect-error: access private member field
     const randomizer = this.faker._randomizer;
     const real = randomizer.next();
+    if (multipleOf > 1) {
+      const minMultiple = Math.ceil(effectiveMin / multipleOf) * multipleOf;
+      const maxMultiple = Math.floor(effectiveMax / multipleOf) * multipleOf;
```
Comment (ST-DDT) [inline src/modules/number/index.ts]: After thinking about this for some time. IMO this is conceptually the same as `effectiveMin` and `effectiveMax`.
So I'm wondering whether we should just move the code up there and always/only follow the `multipleOf` branch/don't make it a separate branch.

File: src/modules/number/index.ts [hunk 5]
```diff
@@ -78,9 +87,30 @@ export class NumberModule extends SimpleModuleBase {
...
+      if (maxMultiple < minMultiple) {
+        throw new FakerError(
+          `No suitable integer value between ${min} and ${max} found.`
+        );
+      }
+
+      const delta = (maxMultiple - minMultiple) / multipleOf + 1;
```
Comment (ST-DDT) [inline src/modules/number/index.ts]: IMO we could omit multipling `minMultiple`+`maxMultiple` with `multipleOf`, then we don't have to divide it here and avoid potential float issues.

**Comment by ST-DDT** [comment]
@matthewmayer What is the status on this PR? Can we help you?

**Comment by matthewmayer** [comment]
Yes, I think I need some help implementing your suggestions

File: test/modules/number.spec.ts [hunk 6]
```diff
@@ -66,6 +66,78 @@ describe('number', () => {
...
+                min: 12,
+                max: 37,
+                multipleOf: 10,
+              })
+            )
+          ),
+        ].sort();
+        expect(results).toEqual([20, 30]);
+      });
+      it('provides numbers with a given multipleOf of 10 with inclusive ends', () => {
```
Comment (ST-DDT) [inline test/modules/number.spec.ts]:
```suggestion
      });

      it('provides numbers with a given multipleOf of 10 with inclusive ends', () => {
```

File: test/modules/number.spec.ts [hunk 7]
```diff
@@ -66,6 +66,78 @@ describe('number', () => {
...
+                min: 10,
+                max: 50,
+                multipleOf: 10,
+              })
+            )
+          ),
+        ].sort();
+        expect(results).toEqual([10, 20, 30, 40, 50]);
+      });
+      it('throws for float multipleOf', () => {
```
Comment (ST-DDT) [inline test/modules/number.spec.ts]:
```suggestion
      });

      it('throws for float multipleOf', () => {
```

File: test/modules/number.spec.ts [hunk 8]
```diff
@@ -66,6 +66,78 @@ describe('number', () => {
...
+          min: 0,
+          max: 10,
+          multipleOf: 0.1,
+        };
+
+        expect(() => faker.number.int(input)).toThrow(
+          new FakerError('multipleOf should be an integer.')
+        );
+      });
+      it('throws for negative multipleOf', () => {
```
Comment (ST-DDT) [inline test/modules/number.spec.ts]:
```suggestion
      });

      it('throws for negative multipleOf', () => {
```

File: test/modules/number.spec.ts [hunk 9]
```diff
@@ -66,6 +66,78 @@ describe('number', () => {
...
+          min: -10,
+          max: 10,
+          multipleOf: -1,
+        };
+
+        expect(() => faker.number.int(input)).toThrow(
+          new FakerError('multipleOf should be greater than 0.')
+        );
+      });
+      it('throws for impossible multipleOf', () => {
```
Comment (ST-DDT) [inline test/modules/number.spec.ts]:
```suggestion
      });

      it('throws for impossible multipleOf', () => {
```

File: src/modules/number/index.ts [hunk 10]
```diff
@@ -82,7 +100,9 @@ export class NumberModule extends SimpleModuleBase {
     // @ts-expect-error: access private member field
     const randomizer = this.faker._randomizer;
     const real = randomizer.next();
-    return Math.floor(real * (effectiveMax + 1 - effectiveMin) + effectiveMin);
+    const delta = effectiveMax - effectiveMin + 1; // +1 for inclusive max bounds and even distribution
+
+    return Math.floor(real * delta + effectiveMin) * multipleOf;
```
Comment (ST-DDT) [inline src/modules/number/index.ts]:
```suggestion
    return Math.floor(real * delta + effectiveMin) * multipleOf;
```

**Review by ST-DDT — APPROVED** [review]

**Comment by ST-DDT** [comment]
@matthewmayer Does the change look good to you?

---

Ready for review.

**Review by Shinigami92 — APPROVED** [review]
lets just assume nobody is checking for the explicit error message in a catch block 😏

**Comment by matthewmayer** [comment]
> lets just assume nobody is checking for the explicit error message in a catch block 😏

i think we can assume thats brittle enough to not be considered breaking
https://xkcd.com/1172/

**Comment by matthewmayer** [comment]
> @matthewmayer Does the change look good to you?
> 
> Ready for review.

looks good for me. i can't approve as its my own PR

**Pull request merged**